demos/common/processing/async_pipeline.cpp

```diff
@@ -0,0 +1,217 @@
+/*
+// Copyright (C) 2018-2020 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+*/
+
+#include "async_pipeline.h"
+#include <samples/args_helper.hpp>
+#include <cldnn/cldnn_config.hpp>
+#include <samples/slog.hpp>
+
+using namespace InferenceEngine;
+
+PipelineBase::PipelineBase(ModelBase* modelInstance, const CnnConfig& cnnConfig, InferenceEngine::Core* engine){
```

Comment:
I suggest passing `shard_ptr<InferenceEngine::Core>`. Why do you mix bare ptrs and smart ptrs? Just go with smart ptrs everywhere.

Reply:
FWIW, there are valid use cases for mixing raw and smart pointers. Passing a `shared_ptr` implies that the receiver will share ownership of the object, we might not need that here. If it were up to me, I would just go with the typical demo approach: create the `Core` in `main` and pass it everywhere (as a raw pointer).

Reply:
It's written in function documentation - you may pass pointer or NULL if you want core to be created automatically. So I suggest that if you pass pointer to the core, it should exists all the time pipeline exists. We may use shared_ptr here but it is really not that necessary as Core should exists all the time that program works - otherwise, it's meaningless.

Reply:
Even though IE doesn't state this, it was required that `ExecutableNetwork` lives as long as `InferRequest`. May be the same was true for the pair `Core` - `ExecutableNetwork`. That would require us to store `shared_ptr<Core>`. I don't know if it is true now. Given that it works for now, there is a chance we don't need this. In this case I'm fine with raw ptr.